### ISSUE TITLE
Sampler for discrete distributions

### ISSUE THREAD

**Issue by ghost** [brian-j-smith/Mamba.jl#11]
The samplers currently available work for continuous distributions, but I think there is not a sampler for discrete distributions. 

I want to test a Bayesian variable selection algorithm that requires a Bernoulli distribution as a prior at the bottom of the hierarchical model.

As far as I know, Mamba [allows](http://mambajl.readthedocs.org/en/latest/tutorial.html#sampling-schemes) you to create your own sampler, although it seems to be focused on conditional distributions for Gibbs sampling. I'm not sure if this can be used to create arbitrary samplers.

**Comment by binarybana** [comment]
@goedman, This is definitely off-topic here, but I've actually implemented another of Dr. Liang's MCMC techniques in Julia [here (SAMC)](https://github.com/binarybana/SAMC.jl). 

I evaluated adding this method to Mamba, but it's sufficiently different from most MCMC techniques that (at the time) it made more sense to implement it by itself (along with MH, and AMWG for comparison (the later being inspired by Mamba's implementation!)).

**Comment by brian-j-smith** [comment]
@binarybana, nice job on the SAMC julia package.

**Comment by goedman** [comment]
@binarybana Hi Jason, thanks for your update/contribution.

I tried to get your README example to work but ran into a couple of issues. As I am working on OS X,

 I can't have 2 files named samc.jl and SAMC.jl. I renamed samc.jl to same_record.jl and updated SAMC.jl accordingly (the 'include' statement).

I stored the provided README example in a little script 'readme.jl' and tried to run it:

'''
julia> include("/Users/rob/Projects/Julia/Rob/SAMC/Examples/readme.jl")
ERROR: `MySampler` has no method matching MySampler(::Float64, ::Float64)
 in include at /Applications/Julia-0.3.4.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include_from_node1 at /Applications/Julia-0.3.4.app/Contents/Resources/julia/lib/julia/sys.dylib
while loading /Users/rob/Projects/Julia/Rob/SAMC/Examples/readme.jl, in expression starting on line 34
'''

Shouldn't the data matrix being added as an argument?

I tried giving it a dummy matrix, like rand(5,5), that makes it fail 2 lines further down:

'''
julia> mh
MHRecord(MySampler(0.0,0.0,5x5 Array{Float64,2}:
 0.567589   0.858875  0.604416   0.0232634  0.502679
 0.0762462  0.573267  0.515966   0.145528   0.907228
 0.479726   0.910614  0.0764755  0.99086    0.792443
 0.0435808  0.166115  0.0707901  0.933435   0.966226
 0.168827   0.872621  0.556068   0.993351   0.611206),MySampler(0.0,0.0,5x5 Array{Float64,2}:
 0.567589   0.858875  0.604416   0.0232634  0.502679
 0.0762462  0.573267  0.515966   0.145528   0.907228
 0.479726   0.910614  0.0764755  0.99086    0.792443
 0.0435808  0.166115  0.0707901  0.933435   0.966226
 0.168827   0.872621  0.556068   0.993351   0.611206),Inf,{},0,0,1,0,1,Dict{Int64,Int64}(),Dict{Int64,Int64}())

julia> include("/Users/rob/Projects/Julia/Rob/SAMC/Examples/readme.jl")
ERROR: unrecognized keyword argument "burn"
 in include at /Applications/Julia-0.3.4.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include_from_node1 at /Applications/Julia-0.3.4.app/Contents/Resources/julia/lib/julia/sys.dylib
while loading /Users/rob/Projects/Julia/Rob/SAMC/Examples/readme.jl, in expression starting on line 36
'''

I'm probably on a completely wrong track here :-)

**Comment by binarybana** [comment]
@goedman , thanks for the bug report! All those are mistakes on my part. I've since cleaned up the API, and made the README and tests the same to make sure they run properly (although proper doctest/literate programming support would make my copy/pasting unnecessary but I digress). So please try again and file bug reports/feature requests on the SAMC tracker. You can now wear your "First official user of the SAMC.jl" badge with pride!

And sorry for the repeated hijacking of this thread!

**Comment by goedman** [comment]
Brian,

Looks like I totally missed this, but DGS() looks like what you described above?

The Eyes example works fine, but I have been trying to warp it into below (Jags) example from Wagenmaker's BCM book:

# Inferring Return Rate and Number of Surveys from Observed Returns
model{
  # Observed Returns
  for (i in 1:m){
     k[i] ~ dbin(theta,n)
  }   
  # Priors on Rate Theta and Number n
  theta ~ dbeta(1,1)
  n ~ dcat(p[])
  for (i in 1:nmax){
    p[i] <- 1/nmax
  }
}   

I believe that should be possible with the DGS() sampler for :p, but I can't get it to work. Any change, if this is possible in Mamba, you can give me a little push in the right direction?

**Comment by brian-j-smith** [comment]
@goedman: Yes, `DSG` is the discrete sampler describe in my earlier post (10/22).  A quick  tip for implementing the Jags example...  the logical nodes are k (the response), theta, and n.  Use `DSG` to sample n and one of the continuous samplers for theta.

**Comment by goedman** [comment]
@brian-j-smith Thanks a lot Brian!

**Comment by brian-j-smith** [comment]
**Long-Term Update:**

In addition to the [Discrete Gibbs Sampler](http://mambajl.readthedocs.org/en/latest/samplers/dgs.html), several binary samplers have been implemented recently in Mamba by @bdeonovic, including:

* [Binary Hamiltonian Monte Carlo (BHMC)](http://mambajl.readthedocs.org/en/latest/samplers/bhmc.html)
* [Binary Metropolised Gibbs (BMG)](http://mambajl.readthedocs.org/en/latest/samplers/bmg.html)
* [Binary Modified Metropolised Gibbs (BMMG)](http://mambajl.readthedocs.org/en/latest/samplers/bmmg.html)

**Comment by rsmith31415** [comment]
@brian-j-smith Great news! Thank you for the update.
